### PR TITLE
JENKINS-62739 Remove Bootstrap CSS

### PR DESCRIPTION
**Pull request by aheritier**
This PR is upgrading  jenkins version requirement from 2.138.4 to 2.204.6 + remove the BootStrap CSS to use the native ones proposed by Jenkins

It fixes #61 #62 #63 #64 #65 

@fqueiruga I wasn't able to use CSS variables like you proposed in https://issues.jenkins-ci.org/browse/JENKINS-62739 because (at least with 2.204.6) it overrides by empty the first value. It gives:

Screenshot 2020-07-20 16 04 16

# Changes:

* JENKINS-62739 Remove Bootstrap CSS
* Bump jenkins version requirement from 2.138.4 to 2.204.6
* Bump hamcrest-library from 1.3 to 2.2
* Bump plugin from 3.50 to 4.4
* Bump gson from 2.8.1 to 2.8.6
* Bump hamcrest-core from 1.3 to 2.2
* Bump support-core from 2.63 to 2.69

# Result

## Configuration reminder

Screenshot 2020-07-20 16 01 44

## Configuration to do

Screenshot 2020-07-20 16 02 17

## Configuration done

Screenshot 2020-07-20 16 07 20

## With 2.245 and Dark Theme

It looks ok with the Dark Theme

FYI @timja 

Screenshot 2020-07-20 16 26 53

Screenshot 2020-07-20 16 27 19

Screenshot 2020-07-20 16 28 28

### CODE REVIEW

File: src/main/resources/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration/index.jelly [hunk 1]
```diff
@@ -4,134 +4,140 @@
   <l:layout title="${it.actionTitleText}" norefresh="true" permission="${it.ADMINISTRATOR}">
 
     <l:header>
-      <link rel="stylesheet" href="${resURL}/bootstrap/css/bootstrap.min.css" type="text/css"/>
       <link rel="stylesheet" href="${resURL}/css/font-awesome/css/font-awesome.min.css"/>
       <style type="text/css">
-        table {
-        border-collapse: separate;
-        border-spacing: 5px;
+        .jenkins-health-advisor .alert-success {
```
Comment (timja) [inline src/main/resources/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration/index.jelly:9]:
```suggestion
        /* TODO remove when baseline newer than 2.245, https://github.com/jenkinsci/jenkins/pull/4842 */
        .jenkins-health-advisor .alert-success {
```

**Comment by aheritier** [comment]
I don't understand why the build is failing on J11. It seems we are passing in the command line `-Djenkins.version=2.164.1` which is the root cause of the error because the support-core plugin and this plugin are now based on 2.204.6.

Not sure where thus parameter is coming from. Maybe @jglick or @oleg-nenashev know ?

File: src/main/resources/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration/index.jelly (hunk 1 above)
Comment (aheritier) [reply src/main/resources/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration/index.jelly:9]: JENKINS-62739 Add a TODO to remove the custom style later

**Comment by timja** [comment]
https://github.com/jenkinsci/cloudbees-jenkins-advisor-plugin/blob/master/Jenkinsfile

Change it to be just:
```
buildPlugin()
```

FTR:
https://github.com/jenkins-infra/pipeline-library/pull/145

**Comment by aheritier** [comment]
@timja 👍  (We really have to keep an eye everywhere)

**Comment by timja** [comment]
https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fcloudbees-jenkins-advisor-plugin/detail/PR-60/5/tests

likely needs a test adjustment to allow files to be deleted properly on windows, there were some changes in jenkins test harness to surface these errors better I think

**Comment by aheritier** [comment]
OMG ... Windows.. Why do we still support this crappy stuff
I will look at it tomorrow, I already spent more thime than expected on it today.
Thanks a lot for your help @timja

**Comment by aheritier** [comment]
it looks to be related to https://github.com/jenkinsci/jenkins-test-harness/pull/185
I have no idea why all these files are still locked.
Not sure what to do.
@jtnord @jglick is there a good practice to implement or fix to do to have a build more windows friendly ?

**Comment by jglick** [comment]
Also https://github.com/jenkinsci/jenkins-test-harness/pull/198 perhaps. What is the problem you are seeing? Maybe some bug in your tests?

**Comment by jglick** [comment]
Not sure what https://github.com/jenkinsci/cloudbees-jenkins-advisor-plugin/blob/2ee22324cdbe25c421854d78ca111a37fd8d3ce7/src/test/java/com/cloudbees/jenkins/plugins/advisor/BundleUploadTest.java#L94-L104 is doing but might need to be suppressed on Windows unless you can track down a specific problem.

Anyway why do you need mocks in https://github.com/jenkinsci/cloudbees-jenkins-advisor-plugin/blob/2ee22324cdbe25c421854d78ca111a37fd8d3ce7/src/test/java/com/cloudbees/jenkins/plugins/advisor/BundleUploadTest.java#L162? You could just use `@LocalData` with a `*.jpi.disabled` file.

**Comment by aheritier** [comment]
@jglick I didn't write this code. I will try to understand better its purpose

AFAIR `execute_pluginDisabled` was written to verify that nothing is sent to Advisor server when the plugin is disabled (I remember some bugs at the origin of the plugin which was sending a bundle when the plugin wasn't yet configured. But I am not sure about the interest of doing such test when the plugin is disabled)

For https://github.com/jenkinsci/cloudbees-jenkins-advisor-plugin/blob/2ee22324cdbe25c421854d78ca111a37fd8d3ce7/src/test/java/com/cloudbees/jenkins/plugins/advisor/BundleUploadTest.java#L162 no idea at all

**Comment by aheritier** [comment]
@jglick is the best (like often) !
I updated/simplified the test to use a "@LocalData" to disable the plugin and windows prefers it.

**Comment by jtnord** [comment]
the standard test plugin manager does not support disabling plugins at run time.  this is because it is reused across tests for speed and does not close the jars that the plugin depends on.

**Comment by aheritier** [comment]
@jtnord great to know. Thanks

File: src/main/resources/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration/index.jelly [hunk 2]
```diff
@@ -4,134 +4,141 @@
...
       <link rel="stylesheet" href="${resURL}/css/font-awesome/css/font-awesome.min.css"/>
       <style type="text/css">
-        table {
-        border-collapse: separate;
-        border-spacing: 5px;
+        /* TODO remove when baseline newer than 2.245, https://github.com/jenkinsci/jenkins/pull/4842 */
+        .jenkins-health-advisor .alert-success {
+          background-color: #d4edda;
+          border-color: #c3e6cb;
+          color: #138347;
```
Comment (fqueiruga) [inline src/main/resources/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration/index.jelly:13]: This should work for theming, right @timja ?

```suggestion
          background-color: #d4edda;
          background-color: var(--alert-success-bg-color);
          border-color: #c3e6cb;
          border-color: var(--alert-success-border-color);
          color: #138347;
          color: var(--alert-success-text-color);
```

File: src/main/resources/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration/index.jelly [hunk 3]
```diff
@@ -4,134 +4,141 @@
...
+        }
+        .jenkins-health-advisor .alert-success p {
+          color: #138347;
```
Comment (fqueiruga) [inline src/main/resources/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration/index.jelly:16]:
```suggestion
          color: #138347;
          color: var(--alert-success-text-color);
```

**Comment by fqueiruga** [comment]
Suggesting changes to allow theming as progressive enhancement with fallback values. I think it should work. (Not a blocker)

**Comment by aheritier** [comment]
@fqueiruga I tried it by copying what you proposed in Jira but it gave me that 👍 



Am I doing something wrong ?

I was test with `mvn hpi:run` thus with 2.204.6

File: src/main/resources/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration/index.jelly (hunk 2 above)
Comment (timja) [reply src/main/resources/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration/index.jelly:13]: yes perfect

**Comment by aheritier** [comment]
Not sure what I did wrong previously @fqueiruga @timja but it's all good with your patch using the CSS variables

# 2.204.6

Screenshot 2020-07-21 12 14 53

# 2.245 + Dark Theme

Screenshot 2020-07-21 12 18 54

It looks good to me

**Comment by fqueiruga** [comment]
To disect the changes a bit,

```less
// Fallback value for IE11
border-color: #c3e6cb;  

// Use the the proper CSS variable and if it is not defined, fall back to using #c3e6cb
// As this declaration is invalid in IE11 it will fall back to the previous one
border-color: var(--alert-success-border-color, #c3e6cb);  
```

**Comment by aheritier** [comment]
ahhh I better understand @fqueiruga 
In https://issues.jenkins-ci.org/browse/JENKINS-62739 you gave me

```
.alert-success {
  background-color: #d4edda;
  background-color: var(--alert-success-bg-color);
  border-color: #c3e6cb;
  border-color: var(--alert-success-border-color);
  color: #138347;
  color: var(--alert-success-color);  
}
```

There was no fallback in the var declaration thus my issue in the rendering and the missing color

> If you haven't I recommend testing it on IE 11 as well :)

AH AH AH IE 11

I think I will do the release and I'll see if I get some bugs reports :-)

**Pull request merged**